apps/s_client.c

```diff
@@ -3034,6 +3034,23 @@ int s_client_main(int argc, char **argv)
      * TODO: Find a better way to do this
      */
     Sleep(50);
+#endif
+    /*
+     * If we ended with an alert being sent, but still with data in the
+     * network buffer to be read, then calling BIO_closesocket() will
+     * result in a TCP-RST being sent. On some platforms (notably
+     * Windows) then this will result in the peer immediately abandoning
+     * the connection including any buffered alert data before it has
+     * had a chance to be read. Shutting down the sending side first,
+     * and then closing the socket sends TCP-FIN first followed by
+     * TCP-RST. This seems to allow the peer to read the alert data.
+     */
+#ifdef _WIN32
+# ifdef SD_SEND
+    shutdown(SSL_get_fd(con), SD_SEND);
+# endif
+#elif defined(SHUT_WR)
+    shutdown(SSL_get_fd(con), SHUT_WR);
 #endif
```

Comment:
You know, names are nice, but I don't think it's appropriate to not call `shutdown()` when a name is missing...  So, errr, you might end up having a call with an explicit 1 in there as well ;-)

Reply:
Well, for windows I am fine to used 1 instead of SD_SEND if that won't work otherwise.
OTOH maybe the ifdef SD_SEND / SHUT_WR is not really a good idea here,
to my excuse I can say I copied it from s_socket.c :)

Reply:
Maybe `s_socket.c` needs a fix as well ,-)

Reply:
So, in general, I could live with using 1 instead of the define.
But does the ifdef here suggest there exist any O/S which actually
fails to define these?

Reply:
Yup, it suggests that Windows might not have `SD_SEND` defined...

Reply:
Also, there might be some old platform that we still claim to build on that doesn't have either, where an explicit 1 is expected...

Reply:
oh, well, I could use something like  `shutdown(x, 1 /* SHUT_WR */)`

Reply:
or
```
#ifndef SHUT_WR
# define SHUT_WR 1
#endif
```
(but I'm late to the party, and it doesn't really matter)